contracts/staking/fuse/StakingRewards.sol

```diff
@@ -0,0 +1,153 @@
+    }
+
+    function earned(address account) public view returns (uint256) {
+        return _pendingStakes[account] * (rewardPerToken() - userRewardPerTokenPaid[account]) / 1e18 + rewards[account];
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 amount) external nonReentrant whenNotPaused {
+        require(amount > 0, "Cannot stake 0");
+        _totalSupply += amount;
+        _pendingStakes[msg.sender] += amount;
+        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
+        emit Staked(msg.sender, amount);
+    }
+
+    function withdraw(uint256 amount) public nonReentrant {
+        require(amount > 0, "Cannot withdraw 0");
+        _totalSupply -= amount;
+        _pendingStakes[msg.sender] -= amount;
+        stakingToken.safeTransfer(msg.sender, amount);
+        emit Withdrawn(msg.sender, amount);
+    }
+
+    function getReward() public nonReentrant {
+        uint256 reward = rewards[msg.sender];
+        if (reward > 0) {
+            rewards[msg.sender] = 0;
+            rewardsToken.safeTransfer(msg.sender, reward);
+            emit RewardPaid(msg.sender, reward);
+        }
+    }
+
+    function exit() external {
+        withdraw(_pendingStakes[msg.sender]);
+        getReward();
+    }
+
+    /* ========== RESTRICTED FUNCTIONS ========== */
+
+    function collectUBIInterest() external {
+        // if (block.timestamp >= periodFinish) {
+        //     rewardRate = reward / rewardsDuration;
+        // } else {
+        //     uint256 remaining = periodFinish - block.timestamp;
+        //     uint256 leftover = remaining * rewardRate;
+        //     rewardRate = (reward + leftover) / rewardsDuration;
+        // }
+        //
+        // // Ensure the provided reward amount is not more than the balance in the contract.
+        // // This keeps the reward rate in the right range, preventing overflows due to
+        // // very high values of rewardRate in the earned and rewardsPerToken functions;
+        // // Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+        // uint balance = rewardsToken.balanceOf(address(this));
+        // require(rewardRate <= balance / rewardsDuration, "Provided reward too high");
+        //
+        // lastUpdateTime = block.timestamp;
+        // periodFinish = block.timestamp + rewardsDuration;
+        // emit RewardAdded(reward);
+
+        // rewardPerTokenStored = rewardPerToken();
+        // lastUpdateTime = lastTimeRewardApplicable();
+        // if (account != address(0)) {
+        //     rewards[account] = earned(account);
+        //     userRewardPerTokenPaid[account] = rewardPerTokenStored;
+        // }
+    }
+
+    // Added to support recovering LP Rewards from other systems such as BAL to be distributed to holders
+    function recoverERC20(address tokenAddress, uint256 tokenAmount) external onlyOwner {
```

Comment:
should be implemented in implementing contract

Reply:
The code was removed.